src/fluree/db/query/fql.cljc

```diff
@@ -11,13 +15,198 @@
             [fluree.db.query.json-ld.response :as json-ld-resp]
             [fluree.db.dbproto :as db-proto]
             [fluree.db.constants :as const])
-  (:refer-clojure :exclude [vswap!])
+  (:refer-clojure :exclude [var? vswap!])
   #?(:cljs (:require-macros [clojure.core])))
 
 #?(:clj (set! *warn-on-reflection* true))
 
 (declare query)
 
+(s/def ::limit pos-int?)
+
+(s/def ::offset nat-int?)
+
+(s/def ::fuel number?)
+
+(s/def ::depth nat-int?)
+
+(s/def ::prettyPrint boolean?)
+
+(s/def ::parseJSON boolean?)
+
+(s/def ::opts (s/keys :opt-un [::parseJSON ::prettyPrint]))
+
+(defn fn-string?
+  [x]
+  (and (string? x)
+       (re-matches #"^\(.+\)$" x)))
+
+(defn fn-list?
+  [x]
+  (and (list? x)
+       (-> x first symbol?)))
+
+(s/def ::function (s/or :string fn-string?, :list fn-list?))
+
+(s/def ::filter (s/coll-of ::function))
+
+(defn wildcard?
+  [x]
+  (#{"*" :* '*} x))
```

Comment:
This isn't something we need to fix right now but I'm curious if there's a way to translate all queries/transactions into a single canonical representation so we don't have to match all three styles.

Reply:
yep, that's in the plan. There's a `normalize` fn here that's called before validation. Eventually, when all the specs are complete, we'll do all the necessary `->kebab-cased-keyword` translation among other things as a first step before validation. The tension here is that I would also like to validate queries on the client before they're sent to the query peer as well, so we need to make sure we can handle all the cases that arise from user input in a clean and unsurprising way as well.